src/NUnitConsole/nunit-console/Utilities/TeamCityServiceMessagePublisher.cs

```diff
@@ -0,0 +1,389 @@
+﻿// ***********************************************************************
+// Copyright (c) 2015 Charlie Poole
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System.IO;
+using System.Xml;
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace NUnit.ConsoleRunner.Utilities
+{
+    using System;
+
+    internal class TeamCityServiceMessagePublisher
+    {
+        private readonly TextWriter _outWriter;
+        private readonly Dictionary<string, string> _refs = new Dictionary<string, string>();
+        private int _blockCounter;
+        private string _rootFlowId;
+
+        public TeamCityServiceMessagePublisher(TextWriter outWriter)
+        {
+            if (outWriter == null)
+            {
+                throw new ArgumentNullException("outWriter");
+            }
+
+            _outWriter = outWriter;
+        }
+
+        public void RegisterMessage(XmlNode message)
+        {
+            if (message == null)
+            {
+                throw new ArgumentNullException("message");
+            }
+
+            var messageName = message.Name;
+            if (string.IsNullOrEmpty(messageName))
+            {
+                return;
+            }
+
+            messageName = messageName.ToLowerInvariant();
+            if (messageName == "start-run")
+            {
+                _refs.Clear();
+                return;
+            }           
+
+            var fullName = message.GetAttribute("fullname");
+            if (string.IsNullOrEmpty(fullName))
+            {
+                return;
+            }
+
+            var id = message.GetAttribute("id");
+            var parentId = message.GetAttribute("parentId");
```

Comment:
- The build log has the hierarchical structure. It is very useful in comparison with AppVeyor and Travis :) Also we has a list of tests, but each test should has an assembly name to distinguish between tests with the same name in different assemblies (for different framework version, different platforms and so on)

- We have a vexing issue related to the hierarchy. If NUnit UI just shows static tree of tests I think we only need to ensure the correctness of these static data, but the tests can have a long run time and the user may want to see the process in real time.
Here we must not forget that the tests can be performed in parallel and we should take into account in the design of application and API. If this feature is not needed, it's all very easy. The TeamCity has the feature to show hierarchical build log in real time.